### PR TITLE
[KOGITO-5299] - Addons Restructure

### PR DESCRIPTION
**Pull request by ricardozanini**
Many thanks for submitting your Pull Request :heart:! 

See: https://issues.redhat.com/browse/KOGITO-5299

In this PR we are restructuring the addons module.

Related PRs: 
https://github.com/kiegroup/kogito-examples/pull/731
https://github.com/kiegroup/kogito-apps/pull/889
https://github.com/kiegroup/kie-docs/pull/3712

TODO:
- [x] Documentation for the main module and the whole concept
- [x] Update Apps, Examples
- [x] Update Docs 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.md)
- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/master/kogito-build/kogito-ide-config)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>Run all builds</b>  
  Please add comment: <b>Jenkins retest this</b>

* <b>Run (or rerun) specific test(s)</b>  
  Please add comment: <b>Jenkins (re)run [runtimes|optaplanner|apps|examples] tests</b>
 
* <b>Quarkus LTS checks</b>  
  Please add comment: <b>Jenkins run LTS</b>

* <b>Run (or rerun) LTS specific test(s)</b>  
  Please add comment: <b>Jenkins (re)run [runtimes|optaplanner|apps|examples] LTS</b>

* <b>Native checks</b>  
  Please add comment: <b>Jenkins run native</b>

* <b>Run (or rerun) native specific test(s)</b>  
  Please add comment: <b>Jenkins (re)run [runtimes|optaplanner|apps|examples] native</b>

* <b>Full Kogito testing</b> (with cloud images and operator BDD testing)  
  Please add comment: <b>Jenkins run BDD</b>  
  <b>This check should be used only if a big change is done as it takes time to run, need resources and one full BDD tests check can be done at a time ...</b>
</details>

### CODE REVIEW

**Comment by ricardozanini** [comment]
Jenkins rerun apps tests

 Jenkins rerun examples tests

**Comment by ricardozanini** [comment]
Apps CI is failing because it's taking the wrong BOM file.

File: addons/README.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+### Add a new Capability to Kogito
+
+In this case, you will have to create a new capability from the scratch. 
+To do that, be aware that you should have knowledge of the Kogito engine internals. Then follow these steps:
+
+1. Create a new sub-module under `api` with a meaningful name. 
+   This module can be a parent module if you foresee a complex scenario for your new add-on. 
+   See [`cloudevents`](api/cloudevents) as an example.
+2. Do not use any dependencies from Quarkus or SpringBoot.
```
Comment (danielezonca) [inline addons/README.md:65]:
```suggestion
2. Do not use any dependencies from Quarkus or SpringBoot in `api` module. 
```

File: addons/README.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+In this case, you will have to create a new capability from the scratch. 
+To do that, be aware that you should have knowledge of the Kogito engine internals. Then follow these steps:
+
+1. Create a new sub-module under `api` with a meaningful name. 
+   This module can be a parent module if you foresee a complex scenario for your new add-on. 
+   See [`cloudevents`](api/cloudevents) as an example.
+2. Do not use any dependencies from Quarkus or SpringBoot. 
+   Your add-on must only have code to support your capability.
+   New dependencies must be added to the [kogito-build](../kogito-build/kogito-build-parent) BOM.
+3. Create the same capability under the runtime you wish to add support (either `quarkus` or `springboot`).
```
Comment (danielezonca) [inline addons/README.md:68]: We need in a future ticket to document how to add `quarkus` component to `kogito-quarkus-bom` and the same for `springboot`.

File: addons/README.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+To do that, be aware that you should have knowledge of the Kogito engine internals. Then follow these steps:
+
+1. Create a new sub-module under `api` with a meaningful name. 
+   This module can be a parent module if you foresee a complex scenario for your new add-on. 
+   See [`cloudevents`](api/cloudevents) as an example.
+2. Do not use any dependencies from Quarkus or SpringBoot. 
+   Your add-on must only have code to support your capability.
+   New dependencies must be added to the [kogito-build](../kogito-build/kogito-build-parent) BOM.
+3. Create the same capability under the runtime you wish to add support (either `quarkus` or `springboot`). 
+   You can choose to give support to only one of them, just make it clear why and discuss this decision with the community.
```
Comment (danielezonca) [inline addons/README.md:69]: Add a note about import `quarkus-bom` or `springboot-dependencies` in these platform specific modules to make sure to use dependencies with the proper version (i.e. supported Kafka)

File: addons/api/pom.xml [hunk 4]
```diff
@@ -3,19 +3,32 @@
...
-    <module>task-management-quarkus-addon</module>
-    <module>task-management-springboot-addon</module>
+    <module>cloudevents</module>
+    <module>events</module>
+    <module>explainability</module>
+    <module>jobs</module>
+    <module>mail</module>
+    <module>monitoring</module>
+    <module>persistence</module>
+    <module>predictions</module>
```
Comment (danielezonca) [inline addons/api/pom.xml:26]: I would like to leverage this PR that will break existing addon GAVs to rename this addon (so here and also nested modules). I find it is very misleading because it looks generic while it is an API specific to predict Human Task data.
Can you please rename like `process-task-predictions` or similar?

@cristianonicolai @tiagodolphine Wdyt? Are you ok with this rename?

File: addons/README.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+Notice that **every** Kogito Add-on starts with `kogito-addons-` prefix. This facilitates your search using IDE:
+
+![](../docsimg/add-on-ide.gif)
+
+In general, the add-on name is composed like `kogito-addons-{runtime}-{capability}-{implementation}`:
+
+- `{runtime}` - can be either `quarkus` or `springboot`.
+- `{capability}` - the capability of the add-on such as `persistence`, `eventing`, `cloudevents` and so forth.
+- `{implementation}` - (optional) the specific implementation of the given capability. For example, `persistence` can have `mongodb` or `postgresql`.
```
Comment (danielezonca) [inline addons/README.md:47]: This classification is nice but we already have examples where it doesn't apply well: `task-management` and others has the `capability` that contains `-` so it clashes with `implementation`.
Personally I'm not sure I like to "formalise" the `capability`/`implementation` classification, probably I would stay with a generic `capability` all the rest

File: addons/quarkus/cloudevents/pom.xml [hunk 6]
```diff
@@ -4,17 +4,19 @@
...
+  <name>Kogito :: Add-Ons :: Quarkus :: Cloud Events</name>
+
+  <description>Kogito Cloud Events</description>
+
   <modules>
-    <module>monitoring-prometheus-common</module>
-    <module>monitoring-prometheus-quarkus-addon</module>
-    <module>monitoring-prometheus-springboot-addon</module>
+    <module>common</module>
+    <module>impl</module>
```
Comment (danielezonca) [inline addons/quarkus/cloudevents/pom.xml:18]: `impl`?

File: addons/quarkus/persistence/infinispan-health/pom.xml [hunk 7]
```diff
@@ -19,13 +19,14 @@
...
   <parent>
-    <artifactId>persistence</artifactId>
+    <artifactId>kogito-addons-quarkus-persistence-parent</artifactId>
     <groupId>org.kie.kogito</groupId>
     <version>2.0.0-SNAPSHOT</version>
   </parent>
   <modelVersion>4.0.0</modelVersion>
 
-  <artifactId>infinispan-quarkus-health-addon</artifactId>
+  <!-- TODO: Since now we have an addon per Runtime, this should be "infinispan" only -->
```
Comment (danielezonca) [inline addons/quarkus/persistence/infinispan-health/pom.xml:28]: What do you mean? Do you mean `quarkus` only instead of `infinispan`?

File: addons/quarkus/pom.xml [hunk 8]
```diff
@@ -29,4 +41,6 @@
       </dependency>
     </dependencies>
   </dependencyManagement>
+
```
Comment (danielezonca) [inline addons/quarkus/pom.xml:44]: This is the new root for all `quarkus` addon so it should be safe to import `quarkus-bom` here

File: addons/springboot/pom.xml [hunk 9]
```diff
@@ -52,4 +47,5 @@
       </dependency>
     </dependencies>
   </dependencyManagement>
+
```
Comment (danielezonca) [inline addons/springboot/pom.xml:50]: This is the root module for `spring` so it should be safe to import `spring-boot-dependencies`

File: kogito-bom/pom.xml [hunk 10]
```diff
@@ -67,85 +67,98 @@
...
         <groupId>org.kie.kogito</groupId>
-        <artifactId>kogito-cloudevents-quarkus-addon</artifactId>
+        <artifactId>kogito-addons-quarkus-cloudevents</artifactId>
         <version>${project.version}</version>
         <classifier>sources</classifier>
       </dependency>
       <dependency>
         <groupId>org.kie.kogito</groupId>
-        <artifactId>kogito-cloudevents-quarkus-addon</artifactId>
+        <artifactId>kogito-addons-quarkus-cloudevents</artifactId>
```
Comment (danielezonca) [inline kogito-bom/pom.xml:117]: I would like to move all `quarkus` modules to (`kogito-quarkus-bom`](https://github.com/kiegroup/kogito-runtimes/blob/master/kogito-quarkus-parent/bom/pom.xml) but we might prefer to do it with another PR. Wdyt?

File: kogito-bom/pom.xml [hunk 11]
```diff
@@ -67,85 +67,98 @@
...
+        <artifactId>kogito-addons-quarkus-cloudevents-multi</artifactId>
         <version>${project.version}</version>
         <classifier>sources</classifier>
       </dependency>
 
 
       <dependency>
         <groupId>org.kie.kogito</groupId>
-        <artifactId>kogito-cloudevents-spring-boot-addon</artifactId>
+        <artifactId>kogito-addons-springboot-cloudevents</artifactId>
```
Comment (danielezonca) [inline kogito-bom/pom.xml:150]: I would like to move all `spring` modules to (`kogito-springboot`](https://github.com/kiegroup/kogito-runtimes/blob/master/kogito-springboot/pom.xml) but we might prefer to do it with another PR. Wdyt?

File: addons/README.md [hunk 12]
```diff
@@ -0,0 +1,88 @@
+# Kogito Add-Ons
+
+In this package you will find the modules responsible to extend the Kogito Core capabilities.
+They are responsible to give persistence, monitoring, messaging, and many other features on top of a plain Kogito service.
+
+## Modules organization
+
+At the root of this project you will find three sub-modules:
+
+1. `api` - contains the core features for a specific add-on, not tied to a particular runtime
```
[the remark this answers is not in the document]
Comment (fjtirado) [reply addons/README.md:10]: I agree with Daniele, Api is misleading, should be common

[the remark this answers is not in the document]
Comment (ricardozanini) [reply addons/README.md:10]: It was `common` at first, but I figured our APIs would be created in the near future for each addon. That's why I've changed. But I'm keen to move back. :shrug:

File: addons/api/pom.xml (hunk 4 above)
Comment (ricardozanini) [reply addons/api/pom.xml:26]: Sure! Please let me know the preferred naming. Renaming these modules takes a lot of effort. I'd rather do it once. :wink:

File: addons/README.md (hunk 5 above)
Comment (ricardozanini) [reply addons/README.md:47]: Yes, indeed, you have a point. But I believe the `-` in the middle of `task-management` or `jobs-service` is pretty self-explained. But sure, it can confuse people. How about a note in this doc clarifying these composed names?

File: addons/api/pom.xml [hunk 13]
```diff
@@ -3,19 +3,32 @@
...
+    <module>process-management</module>
+    <module>process-svg</module>
+    <module>rest-exception-handler</module>
```
Comment (fjtirado) [inline addons/api/pom.xml:29]: Although rest-exception-handler is only used in addons (task-management and process-management) it is not an addon per se. Maybe should be moved to runtimes core.

File: addons/quarkus/cloudevents/pom.xml (hunk 6 above)
Comment (ricardozanini) [reply addons/quarkus/cloudevents/pom.xml:18]: Ouch! iIt means `implementation`. I forgot to change. :facepalm: Actually, I'm open to suggestions to this module name.

File: addons/quarkus/persistence/infinispan-health/pom.xml (hunk 7 above)
Comment (ricardozanini) [reply addons/quarkus/persistence/infinispan-health/pom.xml:28]: One for each Runtime. I should've added the JIRA in this description: https://issues.redhat.com/browse/KOGITO-5325

File: kogito-bom/pom.xml (hunk 10 above)
Comment (ricardozanini) [reply kogito-bom/pom.xml:117]: Yeah, I'd rather do this in a new PR. IF we are going to do this. Not sure if we want? This way I would then refactor as:

```
- addons
-- {capabilities}
- kogito-quarkus
-- extensions
-- addons
--- {capabilities}
- kogito-springboot
-- starters
-- addons
```

?

File: kogito-bom/pom.xml (hunk 11 above)
Comment (ricardozanini) [reply kogito-bom/pom.xml:150]: See my comment above. Yes, I agree. But in another PR.

File: addons/api/pom.xml (hunk 13 above)
Comment (ricardozanini) [reply addons/api/pom.xml:29]: This is an interesting use case. It's not an addon but a shared library. I agree that we might need to move it, but this infrastructure of having a separate handler for each runtime should not be in the core engine. :thinking: 

We can keep it here and document the module intentions or create an `internal` module inside the addons. WDYT?

File: addons/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (ricardozanini) [reply addons/README.md:10]: Renamed parent to `kogito-addons-parent` and folder to `common`.

File: addons/api/pom.xml (hunk 13 above)
Comment (fjtirado) [reply addons/api/pom.xml:29]: If you ask me, I will created a shared module under kogito-runtimes root and put it there.
But since we have a lot of top level directories (although you created a jira to get rid of some ot them like docsimg, history...), for now I will just create a shared dir for addons.
My concerns is that someone can believe this is an addon, maybe, as you propose, a comment might be sufficient. 
I let it up to you (shared or comment)

File: addons/README.md (hunk 2 above)
Comment (ricardozanini) [reply addons/README.md:68]: https://issues.redhat.com/browse/KOGITO-5355

File: addons/api/pom.xml (hunk 4 above)
Comment (ricardozanini) [reply addons/api/pom.xml:26]: @cristianonicolai @tiagodolphine, can you please share your thoughts?

File: addons/README.md (hunk 5 above)
Comment (ricardozanini) [reply addons/README.md:47]: I've added a note in the docs about capabilities with composing names.

File: addons/springboot/pom.xml (hunk 9 above)
Comment (ricardozanini) [reply addons/springboot/pom.xml:50]: It's already in there.

File: kogito-bom/pom.xml (hunk 10 above)
Comment (ricardozanini) [reply kogito-bom/pom.xml:117]: @danielezonca please see https://issues.redhat.com/browse/KOGITO-5356 and add your inputs

File: kogito-bom/pom.xml (hunk 11 above)
Comment (ricardozanini) [reply kogito-bom/pom.xml:150]: See https://issues.redhat.com/browse/KOGITO-5356

**Review by danielezonca — APPROVED** [review]
LGTM 👍

File: addons/api/pom.xml (hunk 4 above)
Comment (danielezonca) [reply addons/api/pom.xml]: I have approved the PR but if possible I would still like to include this renaming. If not possible please create a ticket :)

Comment (ricardozanini) [reply addons/api/pom.xml]: @tiagodolphine could you please reply? This way I can push these changes now and merge EOD.

File: addons/api/pom.xml (hunk 13 above)
Comment (ricardozanini) [reply addons/api/pom.xml]: I'll add this info in the documentation of the `rest-exteception-handler` in this JIRA: https://issues.redhat.com/browse/KOGITO-5300

**Review by fjtirado — APPROVED** [review]

**Review by tiagodolphine — COMMENTED** [review]
LGTM, just left one comment.
Thanks.

File: addons/springboot/monitoring/pom.xml [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+  <parent>
+    <artifactId>kogito-addons-springboot-parent</artifactId>
+    <groupId>org.kie.kogito</groupId>
+    <version>2.0.0-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>kogito-addons-springboot-monitoring-parent</artifactId>
+  <name>Kogito :: Add-Ons :: SpringBoot :: Monitoring :: Parent</name>
+  <description>Kogito Monitoring Addons for Quarkus Runtime</description>
```
Comment (MarianMacik) [inline addons/springboot/monitoring/pom.xml:14]: Should be Spring Boot :)

**Review by MarianMacik — APPROVED** [review]

File: addons/common/cloudevents/pom.xml [hunk 15]
```diff
@@ -21,35 +21,18 @@
   <modelVersion>4.0.0</modelVersion>
   <parent>
     <groupId>org.kie.kogito</groupId>
-    <artifactId>addons</artifactId>
+    <artifactId>kogito-addons-common-parent</artifactId>
     <version>2.0.0-SNAPSHOT</version>
   </parent>
-  <artifactId>cloudevents</artifactId>
+  <artifactId>kogito-addons-cloudevents-parent</artifactId>
```
[the remark this answers is not in the document]
Comment (ricardozanini) [reply addons/common/cloudevents/pom.xml:27]: Oh the `common` module will vanish, see: https://issues.redhat.com/browse/KOGITO-5356

File: addons/springboot/monitoring/pom.xml (hunk 14 above)
Comment (ricardozanini) [reply addons/springboot/monitoring/pom.xml:14]: Great eye!

File: addons/api/pom.xml (hunk 4 above)
Comment (tiagodolphine) [reply addons/api/pom.xml]: Well, TBH I don't have much knowledge about this addon, but I agree if we could be more specific the better it is. What about  `task-prediction`? @danielezonca @ricardozanini

**Review by tiagodolphine — APPROVED** [review]

File: addons/api/pom.xml (hunk 4 above)
Comment (ricardozanini) [reply addons/api/pom.xml]: @danielezonca @cristianonicolai, could you guys please decide on this name? I'll intend to merge it next Monday :)

**Comment by ricardozanini** [comment]
@danielezonca @cristianonicolai I've added a Migration section to the doc. We can add it in the next release notes/announcement.

The only thing left is the naming for the `predictions` module. I like `task-predictions` as @tiagodolphine suggested, but I'm not sure if this name really passes the intentions of the add-on. You tell me. :)

File: addons/api/pom.xml (hunk 4 above)
Comment (cristianonicolai) [reply addons/api/pom.xml]: Im fine with `task-prediction`

**Review by cristianonicolai — APPROVED** [review]

**Comment by cristianonicolai** [comment]
Jenkins rerun runtimes tests

File: addons/api/pom.xml (hunk 4 above)
Comment (ricardozanini) [reply addons/api/pom.xml]: As discussed internally, the name will be `human-task-prediction`.

**Review by r00ta — APPROVED** [review]

**Comment by ricardozanini** [comment]
Jenkins rerun apps tests
Jenkins rerun examples tests

**Comment by ricardozanini** [comment]
Docs updated: https://github.com/kiegroup/kie-docs/pull/3712
